hkmc2/shared/src/main/scala/hkmc2/semantics/Term.scala

```diff
@@ -115,15 +115,29 @@ enum Term extends Statement:
   case Handle(lhs: LocalSymbol, rhs: Term, args: List[Term],
     derivedClsSym: ClassSymbol, defs: Ls[HandlerTermDefinition], body: Term)
   
-  def symbol: Opt[Symbol] = this match
-    case ref: Ref if ref.resSym.nonEmpty => ref.resSym
+  /**
+   * The prelinminary symbol for the term that is resolved during
+   * elaboration. 
+   */
+  lazy val symbol: Opt[Symbol] = this match
     case Ref(sym) => S(sym)
     case sel: Sel => sel.sym
     case sel: SynthSel => sel.sym
     case sel: SelProj => sel.sym
+    case TyApp(lhs = lhs) => lhs.symbol
+    case _ => N
+  
+  /**
+   * The symbol representing the evaluation result of the term. This
+   * symbol is resolved during the resolution stage.
+   */
+  def resolvedSymbol: Opt[Symbol] = this match
+    case ref: Ref => ref.resSym
+    case sel: Sel => sel.sym
+    case sel: SynthSel => sel.sym
+    case sel: SelProj => sel.sym
     case app: App => app.sym
-    case tyApp: TyApp if tyApp.sym.nonEmpty => tyApp.sym
-    case TyApp(prefix, _) => prefix.symbol
+    case tyApp: TyApp => tyApp.sym
     case _ => N
```

Comment:
I restored the `symbol` for the original use case, but kept the case that for `TyApp` inherits the symbol from it LHS because AFAIK applying types to a term never changes its symbol.

I added a new method `resolvedSymbol` that returns the resolved symbol of a term. The places that require symbols after or during resolution should call this method instead. 

For `Sel` they are available from both `symbol` and `resolvedSymbol` because of the reason we discussed in the metting.